### ISSUE TITLE
Add support for processing ddsketch data type as input

### ISSUE THREAD

**Issue by mattklein123** [ClickHouse/ClickHouse#66734]
This is a follow on feature request to https://github.com/ClickHouse/ClickHouse/pull/56342. The current sketch aggregation functions require individual values to be inserted and then merged into the aggregation.

What I would like to be able to do is insert complete sketches (either in binary form or as a custom data type). This way during merging, the database would merge the provided sketch with the aggregated sketch.

I would be happy to implement this feature but I would like some guidance on the best way of going about this, as I think it would require either:
1) A new data type specific to ddsketch (seems unlikely this would be added)
2) More likely a function that takes a binary value and tries to create a quantileDDState from it, which can then be used as input to further aggregated storage. This seems like the best option.

Obviously in order for this to work the binary format of the sketch needs to be agreed upon. I haven't looked yet on the format but it could likely be documented and then parsed as part of the function proposed above.

Let me know if this is an acceptable feature and if so if there are any thoughts on how best to implement. Once that is done I would be happy to implement it.

cc @srikanthccv who did the original feature PR for thoughts.

Thank you for the awesome project!

**Comment by mattklein123** [comment]
Hmm in reviewing the previous PR now I'm second guessing myself that this might already be possible somehow: https://github.com/ClickHouse/ClickHouse/pull/56342/files#diff-46c443dd726a9de237c408e3a82affda8185adc50c54cc5ea325184304cc962bR28-R31. It's not clear to me how one goes about this, unless it as "simple" as just sending the binary and wrapping it in a State()?

**Comment by mattklein123** [comment]
🤦. I think I just answered my own question. I didn't realize it was possible to effectively cast the binary state into and out of the aggregation type, as per https://clickhouse.com/docs/en/sql-reference/data-types/aggregatefunction#data-selection. I will close this and see if I can get it working and report back if I have any issues.

**Comment by srikanthccv** [comment]
Hi @mattklein123, Yes, sending the complete sketches in binary form is possible and was one of the main goals as it would allow storing the pre-aggregated sketches coming from OTEL (known as ExponentialHistogram) or anyone who supports similar log-scale sketches. The binary format is the same as the DDSketch proto. Here are the go client implementations to give an idea https://github.com/SigNoz/ch-go/pull/1, https://github.com/SigNoz/clickhouse-go/pull/1

**Comment by mattklein123** [comment]
> Hi @mattklein123, Yes, sending the complete sketches in binary form is possible and was one of the main goals as it would allow storing the pre-aggregated sketches coming from OTEL (known as ExponentialHistogram) or anyone who supports similar log-scale sketches. The binary format is the same as the DDSketch proto. Here are the go client implementations to give an idea https://github.com/SigNoz/ch-go/pull/1, https://github.com/SigNoz/clickhouse-go/pull/1

Thanks very much. I'm probably being dense but does the format actually correspond to proto encoding explicitly? It doesn't look like https://protobuf.dev/programming-guides/encoding/? Seems like it's missing the field index encoding? I'm not very familiar with proto encoding so I'm probably missing something.

**Comment by mattklein123** [comment]
> Thanks very much. I'm probably being dense but does the format actually correspond to proto encoding explicitly? It doesn't look like https://protobuf.dev/programming-guides/encoding/? Seems like it's missing the field index encoding? I'm not very familiar with proto encoding so I'm probably missing something.

Specifically it looks like the custom encoding used in https://github.com/DataDog/sketches-go/blob/master/ddsketch/encoding/encoding.go which is not the same as the protobuf binary encoding AFAIK, though the Clickhouse side doesn't implement the var float encoding AFAICT so even still it's not exactly the same. I will look through things and try to get it working on my end.

**Comment by srikanthccv** [comment]
You are correct. I didn't mean to say the protobuf encoding but compatibility with sketches libs custom encoding.

**Comment by mattklein123** [comment]
> You are correct. I didn't mean to say the protobuf encoding but compatibility with sketches libs custom encoding.

Thanks I was able to get it working. For anyone else that happens to come along and find this, per above, the binary format is not proto encoding, and it is also not the same as the custom encoding that the various DD libraries use (for example it does not use variable length floats). I will try to do some doc PRs to Clickhouse to make this more clear.

**Comment by nikita-vanyasin** [comment]
Hi @mattklein123 just curious if you still can provide some examples of client which can send binary data to CH? It can be as simple as github Gist - I'll prepare a full example and doc page out of it.
Thank you!

**Comment by mattklein123** [comment]
For the client library I'm using this patch to the Rust client: https://github.com/mattklein123/clickhouse.rs/commit/27aea47bc97cca532ab8a2774ea89e7af68f1949 (asked about upstreaming here https://github.com/ClickHouse/clickhouse-rs/issues/196, looks like there will be some other solution for this.)

For the sketch format the code is here. It's a forked and modified version of the upstream library: https://github.com/mattklein123/sketches-rust/blob/patch-stack/src/store/mod.rs#L31

**Comment by nikita-vanyasin** [comment]
Thanks @mattklein123 
Sending an internal state of aggregate function is certainly not quite reliable, but I see there is value in it. We will discuss with the team if having some kind of 'how-to' guide will be beneficial.
